### PR TITLE
Android tv-casting-app: simplified command API

### PR DESCRIPTION
**Pull request by sharadb-amazon**
This change simplifies and decouples the implementation of the APIs for the LaunchURL command from the old TvCastingApp.java for Android

### Change summary

1.  Added Cluster, MatterCluster, Command, MatterCommand classes which are super classes for the clusters and commands to be supported.
2.  Added new cluster, command objects for Casting use cases in MatterClusters/MattterCommands and the corresponding JNI code int he MatterCommandObjects-JNI.h/.cpp files
3.  Added new fragment to select actions and demonstrate the new LaunchURL API
4.  APIs.md documentation has been updated to illustrate how to use these APIs.

### Testing

Verified and tested locally with Android tv-casting-app running against the Linux TV app

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCluster-JNI.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+JNI_METHOD(jobject, getCommand)
+(JNIEnv * env, jobject thiz, jclass commandClass)
+{
+    chip::DeviceLayer::StackLock lock;
+    ChipLogProgress(AppServer, "MatterCluster-JNI::getCommand() called");
+
+    jclass clsClass        = env->GetObjectClass(commandClass); // Class<Class>
+    jmethodID mid          = env->GetMethodID(clsClass, "getName", "()Ljava/lang/String;");
+    jstring jClassName     = (jstring) env->CallObjectMethod(commandClass, mid);
+    const char * className = env->GetStringUTFChars(jClassName, nullptr);
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCluster-JNI.cpp:59]: please consider to use https://github.com/project-chip/connectedhomeip/blob/2f2c4f1e311c9c35912e21156f816837986bc857/src/lib/support/JniTypeWrappers.h#L29

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCluster-JNI.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    ChipLogProgress(AppServer, "MatterCluster-JNI::getCommand() className: %s", className);
+    if (strcmp(className, "com.matter.casting.clusters.MatterCommands$ContentLauncherClusterLaunchURLCommand") == 0)
+    {
+        BaseCluster * cluster = support::convertClusterFromJavaToCpp(thiz);
+        VerifyOrReturnValue(cluster != nullptr, 0, ChipLogError(AppServer, "MatterCluster-JNI::getCommand() cluster == nullptr"));
+        void * command = cluster->GetCommand(chip::app::Clusters::ContentLauncher::Commands::LaunchURL::Id);
+
+        jobject commandJavaObject = support::convertCommandFromCppToJava(
+            command, "com/matter/casting/clusters/MatterCommands$ContentLauncherClusterLaunchURLCommand");
+        env->ReleaseStringUTFChars(jClassName, className);
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCluster-JNI.cpp:69]: please consider to use https://github.com/project-chip/connectedhomeip/blob/2f2c4f1e311c9c35912e21156f816837986bc857/src/lib/support/JniTypeWrappers.h#L29, when className is null, you should not ReleaseStringUTFChars

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCastingPlayer-JNI.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+#include <jni.h>
+
+namespace matter {
+namespace casting {
+namespace core {
+
+class ConnectionContextJNI
+{
+public:
+    CastingPlayer * castingPlayer;
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCastingPlayer-JNI.h:32]: please init all these values.
maybe struct is better here? considering all these public members.

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCluster-JNI.h [hunk 4]
```diff
@@ -1,6 +1,7 @@
 /*
  *
- *    Copyright (c) 2023 Project CHIP Authors
+ *    Copyright (c) 2020-24 Project CHIP Authors
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCluster-JNI.h:3]: 2020-2024

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/clusters/MatterCommandObjects-JNI.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
+/*
+ *
+ *    Copyright (c) 2020-24 Project CHIP Authors
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/clusters/MatterCommandObjects-JNI.h:3]: 2020-2024

File: examples/tv-casting-app/android/App/app/src/main/java/com/matter/casting/ContentLauncherLaunchURLExampleFragment.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+          MatterCommands.ContentLauncherClusterLaunchURLRequest request =
+              new MatterCommands.ContentLauncherClusterLaunchURLRequest();
+          request.contentURL =
+              ((EditText) getView().findViewById(R.id.contentUrlEditText)).getText().toString();
+          request.displayString =
+              ((EditText) getView().findViewById(R.id.contentDisplayStringEditText))
+                  .getText()
+                  .toString();
+          CompletableFuture<MatterCommands.ContentLauncherClusterResponse> responseFuture =
+              command.invoke(request, 5000);
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/java/com/matter/casting/ContentLauncherLaunchURLExampleFragment.java:121]: I would suggest to use 
https://github.com/project-chip/connectedhomeip/blob/2f2c4f1e311c9c35912e21156f816837986bc857/src/controller/java/src/chip/devicecontroller/ChipDeviceController.java#L1192

AndroidInteractionClient.cpp and its JNI

https://github.com/project-chip/connectedhomeip/blob/2f2c4f1e311c9c35912e21156f816837986bc857/src/controller/java/AndroidInteractionClient.cpp

Comment (yunhanw-google) [reply examples/tv-casting-app/android/App/app/src/main/java/com/matter/casting/ContentLauncherLaunchURLExampleFragment.java:121]: We should try our best to leverage the existing Java/JNI code for InteractionModel and reduce the duplicate maintenance effort.

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/clusters/MatterCommandObjects-JNI.cpp [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        jfieldID dataField = env->GetFieldID(responseTypeClass, "data", "Ljava/lang/String;");
+        char * buffer      = new char[response->data.Value().size() + 1];
+        strncpy(buffer, response->data.Value().data(), response->data.Value().size());
+        buffer[response->data.Value().size()] = '\0';
+        env->SetObjectField(jResponseObj, dataField, env->NewStringUTF(buffer));
+        delete[] buffer;
+    }
+
+    jclass integerClass  = env->FindClass("java/lang/Integer");
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/clusters/MatterCommandObjects-JNI.cpp:84]: why don't you check whether it is null?

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/clusters/MatterCommandObjects-JNI.cpp [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    JNIEnv * env = chip::JniReferences::GetInstance().GetEnvForCurrentThread();
+
+    jclass responseTypeClass = nullptr;
+    VerifyOrReturnValue(
+        chip::JniReferences::GetInstance().GetClassRef(
+            env, "com/matter/casting/clusters/MatterCommands$ContentLauncherClusterResponse", responseTypeClass) == CHIP_NO_ERROR,
+        nullptr, ChipLogError(AppServer, "convertResponseFromCppToJava could not get ContentLauncherClusterResponse class ref"));
+
+    jmethodID constructor = env->GetMethodID(responseTypeClass, "<init>", "()V");
+    jobject jResponseObj  = env->NewObject(responseTypeClass, constructor);
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/clusters/MatterCommandObjects-JNI.cpp:72]: we need check jResponseObj is null or not

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/clusters/MatterCommandObjects-JNI.h [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    static jobject Invoke(jobject jCommand, jobject jRequest, jobject jTimedInvokeTimeoutMs)
+    {
+        const char * commandClassName = GetCommandClassName(jCommand);
+        VerifyOrReturnValue(commandClassName != nullptr, nullptr,
+                            ChipLogError(AppServer, "MatterCommandManager: Could not get commandClassName from jCommand"));
+
+        if (strcmp(commandClassName, "com.matter.casting.clusters.MatterCommands$ContentLauncherClusterLaunchURLCommand") == 0)
+        {
+            auto * command = new clusters::ContentLauncherClusterLaunchURLCommand(jCommand);
+            return command->Invoke(jRequest, jTimedInvokeTimeoutMs);
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/clusters/MatterCommandObjects-JNI.h:69]: what if new fails, then crash happens...

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCommandTemplate-JNI.h [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        VerifyOrReturnValue(GetCppRequestFromJava(jRequest, cppRequest) == CHIP_NO_ERROR, nullptr);
+
+        // call Invoke on launchURLCommand while passing in success/failure callbacks
+        cppCommand->Invoke(
+            cppRequest, nullptr,
+            [completableFuture, this](void * context, const typename Type::ResponseType & response) {
+                ChipLogProgress(AppServer, "<MatterCommandTemplateJNI::invoke()> converting 'response' from Cpp to Java");
+                jobject jResponse = GetJResponseFromCpp(&response);
+
+                JNIEnv * env = chip::JniReferences::GetInstance().GetEnvForCurrentThread();
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCommandTemplate-JNI.h:99]: no env check

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCommandTemplate-JNI.h [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+                chip::DeviceLayer::StackUnlock unlock;
+                env->CallBooleanMethod(completableFuture, mCompleteMethod, jResponse);
+            },
+            [completableFuture, this](void * context, CHIP_ERROR error) {
+                JNIEnv * env         = chip::JniReferences::GetInstance().GetEnvForCurrentThread();
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/MatterCommandTemplate-JNI.h:104]: this is the callback flow, the whole code is leaking local reference...

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/support/Converters-JNI.cpp [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+namespace support {
+
+using namespace chip;
+
+jobject convertMatterErrorFromCppToJava(CHIP_ERROR inErr)
+{
+    JNIEnv * env = chip::JniReferences::GetInstance().GetEnvForCurrentThread();
+    jclass jMatterErrorClass;
+    CHIP_ERROR err =
+        chip::JniReferences::GetInstance().GetClassRef(env, "com/matter/casting/support/MatterError", jMatterErrorClass);
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/support/Converters-JNI.cpp:33]: GetClassRef has been removed from library...

File: examples/tv-casting-app/android/App/app/src/main/jni/cpp/support/Converters-JNI.cpp [hunk 13]
```diff
@@ -0,0 +1,268 @@
...
+
+    jmethodID jMatterErrorConstructor = env->GetMethodID(jMatterErrorClass, "<init>", "(JLjava/lang/String;)V");
+
+    return env->NewObject(jMatterErrorClass, jMatterErrorConstructor, inErr.AsInteger(), nullptr);
+}
+
+jobject convertEndpointFromCppToJava(matter::casting::memory::Strong<core::Endpoint> endpoint)
+{
+    ChipLogProgress(AppServer, "EndpointConverter-JNI.convertEndpointFromCppToJava() called");
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
```
Comment (yunhanw-google) [inline examples/tv-casting-app/android/App/app/src/main/jni/cpp/support/Converters-JNI.cpp:44]: no check on env...

**Review by yunhanw-google — CHANGES_REQUESTED** [review]
Please leverage the existing the Java/Jni invoke API instead of creating new path and adding the additional burden for maintenance..

Please address all leaks.

**Comment by yunhanw-google** [comment]
> Please leverage the existing the Java/Jni invoke API instead of creating new path and adding the additional burden for maintenance..
> 
> Please address all leaks.

@sharadb-amazon 
our controller IM Java/JNI code is couple of deviceControllerPtr, aka deviceProxy. https://github.com/project-chip/connectedhomeip/blob/2f2c4f1e311c9c35912e21156f816837986bc857/src/controller/java/CHIPDeviceController-JNI.cpp#L264, I think we may introduce a deviceProxy function that can be used by Java application which don't need commissioning functionality and need IM functionality?

**Comment by sharadb-amazon** [comment]
Canceling this PR. Will do this work in parts. I have a separate PR out that just exposed the casting Endpoint API on Android and does some GUI work: #32552 Then, will send a separate PR out that exposes the Interaction Model for Android.

**Pull request closed without merging**